src/Interpreters/Cache/FileSegment.cpp

```diff
@@ -513,64 +488,54 @@ void FileSegment::setDownloadFailedUnlocked(std::unique_lock<std::mutex> & segme
 
 void FileSegment::completePartAndResetDownloader()
 {
-    std::unique_lock segment_lock(mutex);
-    completePartAndResetDownloaderUnlocked(segment_lock);
+    auto lock = segment_guard.lock();
+    completePartAndResetDownloaderUnlocked(lock);
 }
 
-void FileSegment::completePartAndResetDownloaderUnlocked(std::unique_lock<std::mutex> & segment_lock)
+void FileSegment::completePartAndResetDownloaderUnlocked(const FileSegmentGuard::Lock & lock)
 {
-    assertNotDetachedUnlocked(segment_lock);
-    assertIsDownloaderUnlocked("completePartAndResetDownloader", segment_lock);
+    assertNotDetachedUnlocked(lock);
+    assertIsDownloaderUnlocked("completePartAndResetDownloader", lock);
 
-    resetDownloadingStateUnlocked(segment_lock);
-    resetDownloaderUnlocked(segment_lock);
+    resetDownloadingStateUnlocked(lock);
+    resetDownloaderUnlocked(lock);
 
-    LOG_TEST(log, "Complete batch. ({})", getInfoForLogUnlocked(segment_lock));
+    LOG_TEST(log, "Complete batch. ({})", getInfoForLogUnlocked(lock));
     cv.notify_all();
 }
 
-void FileSegment::completeWithState(State state)
+void FileSegment::setBroken()
 {
-    std::lock_guard cache_lock(cache->mutex);
-    std::unique_lock segment_lock(mutex);
-
-    assertNotDetachedUnlocked(segment_lock);
-    assertIsDownloaderUnlocked("complete", segment_lock);
-
-    if (state != State::DOWNLOADED
-        && state != State::PARTIALLY_DOWNLOADED
-        && state != State::PARTIALLY_DOWNLOADED_NO_CONTINUATION)
-    {
-        cv.notify_all();
-        throw Exception(
-            ErrorCodes::LOGICAL_ERROR,
-            "Cannot complete file segment with state: {}", stateToString(state));
-    }
-
-    setDownloadState(state);
-    completeBasedOnCurrentState(cache_lock, segment_lock);
+    auto lock = segment_guard.lock();
+    assertNotDetachedUnlocked(lock);
+    assertIsDownloaderUnlocked("setBroken", lock);
+    resetDownloadingStateUnlocked(lock);
+    resetDownloaderUnlocked(lock);
 }
 
-void FileSegment::completeWithoutState()
+void FileSegment::complete()
 {
-    std::lock_guard cache_lock(cache->mutex);
-    completeWithoutStateUnlocked(cache_lock);
+    auto cache_lock = cache->cacheLock();
+    auto locked_key = createLockedKey();
+    return completeUnlocked(*locked_key, cache_lock);
 }
 
-void FileSegment::completeWithoutStateUnlocked(std::lock_guard<std::mutex> & cache_lock)
+void FileSegment::completeUnlocked(LockedKey & locked_key, const CacheGuard::Lock & cache_lock)
 {
-    std::unique_lock segment_lock(mutex);
-    completeBasedOnCurrentState(cache_lock, segment_lock);
-}
+    auto segment_lock = segment_guard.lock();
 
-void FileSegment::completeBasedOnCurrentState(std::lock_guard<std::mutex> & cache_lock, std::unique_lock<std::mutex> & segment_lock)
-{
-    if (is_detached)
+    if (download_state == State::DETACHED)
+    {
+        assertDetachedStatus(segment_lock);
```

Comment:
a strange check which will never trigger error if `download_state == DETACHED`

Reply:
now yes, because of some changes as part of review fixes, but previously it made sense, will remove as now indeed not needed.